### PR TITLE
[2/N] Added the core structure of elastic EP and the eplb algorithm with faulty rank

### PR DESCRIPTION
**Pull request by HanHan009527**
## Motivation

1. Integrating Mooncake's fault-awareness, we need to adjust the eplb algorithm and model loading logic to enable the forward pass to bypass faulty ranks.

base on #10423
check our next pr #11657 and full draft #8961 (update on 10.16) to test the effect of fault redundancy

The ut part is modified to facilitate testing on machines with different ibdev names

## Modifications
1. Adding the core structure of the elastic ep module to bridge forward propagation and scheduling.
2. Add an EPLB algorithm that supports fault rank.

## Accuracy Tests
Test results are available from our full draft version. #8961 

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.ai/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.ai/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.ai/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.ai/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.ai/developer_guide/contribution_guide.html#benchmark-the-speed).

### CODE REVIEW

**Comment by ShangmingCai** [comment]
CC: @zhihui1084

File: python/sglang/test/test_disaggregation_utils.py [hunk 1]
```diff
@@ -100,41 +103,57 @@ def tearDownClass(cls):
...
+    ]
+    logger.info("Resolved rdma_all_devices=%s", rdma_all_devices)
+
+    n_rdma = len(rdma_all_devices)
     # 1. Get visible GPU indices
     cuda_visible_devices = os.getenv("CUDA_VISIBLE_DEVICES")
     if not cuda_visible_devices:
         warnings.warn("CUDA_VISIBLE_DEVICES is not set. Using default RDMA devices.")
-        return "mlx5_roce0,mlx5_roce4"
+        return ",".join(_pick_default_pair(rdma_all_devices))
```
Comment (ShangmingCai) [inline python/sglang/test/test_disaggregation_utils.py]: IB device topo detection and auto fault tolerance feature of mooncake is buggy, I think we only need one IB from each NUMA, it is enough for the current CI.

Comment (HanHan009527) [reply python/sglang/test/test_disaggregation_utils.py]: This part is to enable the volcano h20 to perform the test. The mellanox roce card of that machine is named  to mlx5_1, etc. Let me move this part to another pr

File: python/sglang/srt/elastic_ep/elastic_ep.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+        )
+
+    @classmethod
+    def healthy_rank_state(
+        cls, *, ep_size: Optional[int], device: Optional[torch.device]
+    ) -> torch.Tensor:
+        size = ep_size if ep_size is not None else torch.distributed.get_world_size()
+        dev = device if device is not None else cls._select_device()
+
+        return torch.ones(size, dtype=torch.int32, device=dev)
```
Comment (ShangmingCai) [inline python/sglang/srt/elastic_ep/elastic_ep.py:74]: Should this dtype be changed to` torch.int64` to align with the future usage?

File: python/sglang/srt/elastic_ep/elastic_ep.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+            last_active_ranks=active.clone(),
+            active_ranks_cpu=active.detach().cpu().clone(),
+        )
+
+    @classmethod
+    def healthy_rank_state(
+        cls, *, ep_size: Optional[int], device: Optional[torch.device]
+    ) -> torch.Tensor:
+        size = ep_size if ep_size is not None else torch.distributed.get_world_size()
+        dev = device if device is not None else cls._select_device()
```
Comment (ShangmingCai) [inline python/sglang/srt/elastic_ep/elastic_ep.py:72]: nit: maybe give `ep_size` and `device` a default value: `None`

**Review by ShangmingCai — APPROVED** [review]
LGTM. Changes are clean.

File: python/sglang/srt/elastic_ep/elastic_ep.py (hunk 2 above)
Comment (UNIDY2002) [reply python/sglang/srt/elastic_ep/elastic_ep.py:74]: Mooncake EP currently uses int32. BTW, what does "future usage" refer to?

Comment (ShangmingCai) [reply python/sglang/srt/elastic_ep/elastic_ep.py:74]: @UNIDY2002 Just wonder if it should align with `log2phy`, which is int64. But I am not sure.

Comment (UNIDY2002) [reply python/sglang/srt/elastic_ep/elastic_ep.py:74]: I had a quick verification. `active_ranks` has two usages in `rebalance_experts`: (1) `num_active_ranks = active_ranks.sum().item()`; (2) `active_ranks_list = active_ranks.tolist()`, so I think using int32 for `active_ranks` may be okay.

File: python/sglang/srt/elastic_ep/elastic_ep.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+    _instance: Optional[ElasticEPState] = None
+    _lock = threading.Lock()
+
+    @classmethod
+    def instance(cls) -> ElasticEPState:
+        return cls._instance
+
+    @classmethod
+    def init(cls, server_args: ServerArgs):
+        with cls._lock:
```
Comment (fzyzcjy) [inline python/sglang/srt/elastic_ep/elastic_ep.py]: nit: shall we init it in one single thread and only once, then the code can be simplified here

**Review by fzyzcjy — COMMENTED** [review]
LGTM only a nit

File: python/sglang/srt/elastic_ep/elastic_ep.py (hunk 4 above)
Comment (HanHan009527) [reply python/sglang/srt/elastic_ep/elastic_ep.py]: get, I will remove this lock. This part just follows the customary writing style of singleton, which is originally single-threaded.

File: python/sglang/srt/elastic_ep/elastic_ep.py (hunk 3 above)
Comment (HanHan009527) [reply python/sglang/srt/elastic_ep/elastic_ep.py:72]: done
